### DISCUSSION TITLE
Hyperlinks in dialog boxes

### DISCUSSION THREAD

**Discussion by cgreenland** [winsiderss/systeminformer#1499] [Questions / Answers] [answered]
I'm part of a student project trying to add support for MetaDefender Cloud to System Informer as a compliment to VirusTotal (see [1373](https://github.com/winsiderss/systeminformer/issues/1373)).

I'm trying to find a way to add clickable hyperlinks to a couple of dialog boxes within the OnlineChecks plugin.

The first is a popup dialog for inputting an API key. I've used PhCreateWindow() in a similar way to the network ping tool, which gets the key just fine, but I haven't found a way to put a hyperlink in the dialog text (a link to obtain an API key).

The second is a error dialog shown when exceeding the daily usage limit. I can parse the HTTP 429 response and use RaiseUploadError() to display a message, but haven't found a way to format hyperlinks with that function. I'm now experimenting with CreateWindowEx(). I've been reading the docs on [SysLink Controls](https://learn.microsoft.com/en-us/windows/win32/controls/create-syslink-controls) and here is my attempt to call CreateWindowEx() using their sample code from within UploadCheckThreadStart().

```c
else if (code == 429000)
{
    CreateWindowExW(0, WC_LINK,
    L"For more information, <A HREF=\"https://www.microsoft.com\">click here</A> ",
    WS_VISIBLE | WS_CHILD | WS_TABSTOP,
    0, 0, CW_USEDEFAULT, CW_USEDEFAULT,
    context->DialogHandle, NULL, PluginInstance->DllBase, NULL);
}
```
I'm not sure if I'm calling it correctly, it seems to just hang on the "scanning" dialog. Particularly unsure if I'm using the right handle and instance.

Is there a way to do any of this? An alternative to links would be a button that acts as a link, but I'm unsure if that's any easier. Any help would be much appreciated, this has been stumping me for over a week.

**Comment by dmex** [comment, marked as answer]
You have to use a dialog box or a rich edit control. 

Visual Studio > View > Other Windows > Resource View (opens on right side of screen) -> Expand solution -> Add resource -> Dialog 

Then just drag+drop controls from the editor to the dialog and show the dialog using the DialogBox function:

![image](https://user-images.githubusercontent.com/1306177/203141138-22e10042-a3b4-4c6c-8db2-136ad2f1494a.png)

**Reply by cgreenland** [reply]
Thanks for your reply @dmex. I made a dialog box using the editor, but when I try to call it with DialogBox nothing happens. Not sure if I'm using the right parameters.
```c
DialogBox(PluginInstance->DllBase, MAKEINTRESOURCE(IDD_MDC_LIMIT), context->DialogHandle, LimitDlgProc);
```
I assume I need a processing function for the dialog, so I made a basic one from some sample code. I'm sure this isn't all I need, but I should at least be able to see the dialog box, right?
```c
INT_PTR CALLBACK LimitDlgProc(_In_ HWND hDlg, _In_ UINT message,
    _In_ WPARAM wParam, _In_ LPARAM lParam)
{
    switch (message)
    {
    case WM_INITDIALOG:
        return TRUE;
    case WM_COMMAND:
        switch (LOWORD(wParam))
        {
        case IDOK:
        case IDCANCEL:
            EndDialog(hDlg, 0);
            return TRUE;
        }
        break;
    }
    return FALSE;
}
```

**Reply by cgreenland** [reply]
What control type do you recommend using for hyperlinks in System Informer? I had used an MFC link control for the above. It looks like adding the MFC link control is what's preventing the dialog from displaying. If I add `DS_NOFAILCREATE` the dialog box will display, but without the MFC link. I'm guessing it's not supported or enabled for System Informer.

A SysLink control will display, but I get an error `RC2104	undefined keyword or key name: https://www.example.com\` if I try to use the HREF attribute, even though the [MS docs](https://learn.microsoft.com/en-us/windows/win32/controls/create-syslink-controls) say that's supported.
```c
//OnlineChecks.rc
IDD_MDC_LIMIT DIALOGEX 0, 0, 310, 177
STYLE DS_SETFONT | DS_MODALFRAME | DS_FIXEDSYS | WS_POPUP | WS_CAPTION | WS_SYSMENU | DS_NOFAILCREATE
CAPTION "Dialog"
FONT 8, "MS Shell Dlg", 400, 0, 0x1
BEGIN
    LTEXT           "MetaDefender Cloud API daily usage limit exceeded, retry after the limit is reset or upgrade API key limits.",IDC_STATIC,58,67,188,20
    DEFPUSHBUTTON   "OK",IDOK,252,156,50,14
    CONTROL         "<a href=\"http://www.example.com\">https://www.example.com</a>",IDC_SYSLINK1,"SysLink",WS_TABSTOP,201,92,60,14
END
```

**Reply by cgreenland** [reply]
Found the problem with SysLink, I was using the wrong escape characters for the inner quotes. When I typed the URL tag into the caption attribute from within the Dialog Editor (instead of editing the .rc file directly) it used `""` insted of `\"` and that worked. Subtle.

![2022-11-26 22_17_54-Plugins - Microsoft Visual Studio](https://user-images.githubusercontent.com/57048220/204117728-5b9eab92-e07e-48c1-b2b2-6cea4781280c.png)

```c
//OnlineChecks.rc
IDD_MDC_LIMIT DIALOGEX 0, 0, 310, 177
STYLE DS_SETFONT | DS_MODALFRAME | DS_FIXEDSYS | WS_POPUP | WS_CAPTION | WS_SYSMENU
CAPTION "Dialog"
FONT 8, "MS Shell Dlg", 400, 0, 0x1
BEGIN
    LTEXT           "MetaDefender Cloud API daily usage limit exceeded, retry after the limit is reset or upgrade API key limits.",IDC_STATIC,58,67,188,20
    DEFPUSHBUTTON   "OK",IDOK,252,156,50,14
    CONTROL         "<a href=""http://www.example.com"">https://www.example.com</a>",IDC_SYSLINK1,
                    "SysLink",WS_TABSTOP,57,94,88,14
END
```
My callback code for the SysLink is below for reference (derived from [here](https://learn.microsoft.com/en-us/windows/win32/controls/use-syslink-notifications)). I used `HWND g_hLink = GetDlgItem(hDlg, IDC_SYSLINK1)` for the check `(((LPNMHDR)lParam)->hwndFrom == g_hLink)`. I'm not sure if it's needed, but hopefully it's proper.
```c
INT_PTR CALLBACK LimitDlgProc(_In_ HWND hDlg, _In_ UINT message,
    _In_ WPARAM wParam, _In_ LPARAM lParam)
{
    switch (message)
    {
    case WM_INITDIALOG:
        return TRUE;

    case WM_COMMAND:
        switch (LOWORD(wParam))
        {
        case IDOK:
        case IDCANCEL:
            EndDialog(hDlg, 0);
            return TRUE;
        }
        break;
    case WM_NOTIFY:
        switch (((LPNMHDR)lParam)->code)
        {
        case NM_CLICK:          // Fall through to the next case.
        case NM_RETURN:
            {
                PNMLINK pNMLink = (PNMLINK)lParam;
                LITEM   item = pNMLink->item;
                HWND g_hLink = GetDlgItem(hDlg, IDC_SYSLINK1);
                if (((LPNMHDR)lParam)->hwndFrom == g_hLink && (item.iLink == 0))
                {
                    PhShellExecute(NULL, item.szUrl, NULL);
                }

                break;
            }
        }
        break;
    }
    return FALSE;
}
```

Thank you for pointing me in the right direction @dmex!

**Comment by cgreenland** [comment]
Sorry to keep spamming my own thread, but I have a followup issue. I have my custom error/dialog box appearing over the "Scanning processname..." dialog whenever I get a 429 response (first image), which is fine, but when I close my dialog box the the "Scanning..." dialog remains indefinitely (second image).

How can I graceful have the program close this dialog/stop the scanning process after my dialog is closed? The normal calls to `RaiseUploadError()` somehow seem to do this, but I'm replacing this with `DialogBox()` for the ability to include hyperlinks. Thanks in advance.

![2022-11-27 20_51_06-System Informer](https://user-images.githubusercontent.com/57048220/204175084-0745b0b9-3440-42b1-9e20-7fd0c44e9cd0.png)
![2022-11-27 20_51_57-Scanning acrotray exe](https://user-images.githubusercontent.com/57048220/204175121-3020a8df-0e12-4697-8278-30be72b19636.png)

**Reply by cgreenland** [reply]
Seems like this does the trick:
```c
SendMessage(context->DialogHandle, TDM_CLICK_BUTTON, IDOK, 0);
```

**Reply by dmex** [reply]
>  I have my custom error/dialog box appearing over the "Scanning processname..." dialog whenever I get a 429 response

You can just use TDM_NAVIGATE_PAGE to update the existing dialog with the error message instead of showing multiple overlapping dialogs?
